uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/LeaveBalanceChange.php

```diff
@@ -839,4 +827,179 @@ public static function getTotalTOILBalanceChangeForContact($contactID, $absenceT
+   *
+   * @param \CRM_HRLeaveAndAbsences_BAO_LeaveRequest $leaveRequest
+   *
+   * @return int
+   *   The number of records updated
+   */
+  public static function recalculateExpiredBalanceChangesForLeaveRequestPastDates(LeaveRequest $leaveRequest) {
+    $numberOfRecordsUpdated = 0;
+
+    $expiredBalanceChanges = self::getExpiredBalanceChangesForLeaveRequestPeriod($leaveRequest);
+    $leaveRequestPastDates = self::getDatesOverlappingBalanceChangesToExpire($expiredBalanceChanges);
+
+    if(empty($expiredBalanceChanges) || empty($leaveRequestPastDates)) {
+      return $numberOfRecordsUpdated;
+    }
+
+    return self::recalculateBalanceChange($expiredBalanceChanges, $leaveRequestPastDates);
+  }
+
+  /**
+   * Get the expired TOIL/Brought Forward LeaveBalanceChanges linked to the LeaveRequest Contact
+   * whose expiry date was within the Absence Period linked to the given LeaveRequest
+   * and also the expiry date is greater than the LeaveRequest from_date and the expired LeaveBalanceChange
+   * absence type is same as that on the leave request.
+   *
+   * This is to ensure that only expired TOIL/Brought Forward LeaveBalanceChanges that are eligible for recalculation
+   * are fetched by the query. If the LeaveBalanceChange has expired before the LeaveRequest from_date,
+   * then there is no need to fetch it.
+   *
+   * @param \CRM_HRLeaveAndAbsences_BAO_LeaveRequest $leaveRequest
+   *
+   * @return array
+   */
+  private static function getExpiredBalanceChangesForLeaveRequestPeriod(LeaveRequest $leaveRequest) {
+    $balanceChangeTable = self::getTableName();
+    $toilRequestTable = TOILRequest::getTableName();
+    $leaveRequestTable = LeaveRequest::getTableName();
+    $periodEntitlementTable = LeavePeriodEntitlement::getTableName();
+    $absencePeriodTable = AbsencePeriod::getTableName();
+
+    $leaveRequestFromDate = new DateTime($leaveRequest->from_date);
+    $leaveRequestToDate =  new DateTime($leaveRequest->to_date);
+
+    $absencePeriod = AbsencePeriod::getPeriodContainingDates(
+      $leaveRequestFromDate,
+      $leaveRequestToDate
+    );
+
+    $query = "
+      SELECT 
+        bc.*,
+        coalesce(absence_period.start_date, toil_leave_request.from_date) as start_date
+      FROM {$balanceChangeTable} bc
+      LEFT JOIN {$toilRequestTable} toil_request
+            ON bc.source_type = %1 AND bc.source_id = toil_request.id
+      LEFT JOIN {$leaveRequestTable} toil_leave_request
+            ON toil_request.leave_request_id = toil_leave_request.id
+      LEFT JOIN {$periodEntitlementTable} period_entitlement
+            ON bc.source_type = %2 AND bc.source_id = period_entitlement.id
+      LEFT JOIN {$absencePeriodTable} absence_period
+            ON period_entitlement.period_id = absence_period.id
+      WHERE bc.expiry_date IS NOT NULL AND
+            (bc.expiry_date BETWEEN %3 AND %4) AND
+            bc.expiry_date >= %5 AND  
+            bc.expired_balance_change_id IS NOT NULL AND
```

Comment:
I was under the impression that it is the expiry records we need to fetch for the recalculation and not the original record. Will update accordingly.